src/tags-input.js

```diff
@@ -35,12 +35,18 @@
  *                                                   When this flag is true, addOnEnter, addOnComma, addOnSpace, addOnBlur and
  *                                                   allowLeftoverText values are ignored.
  * @param {boolean=} [spellcheck=true] Flag indicating whether the browser's spellcheck is enabled for the input field or not.
+ * @param {expression} onTagAdding A reference to a method in scope that will be invoked before adding a new tag. This is an option for additional, custom validations or a general way to prevent a tag from being added. The new tag is passed to the function for evaluation. This method must return either true or false. If false, the tag will not be added.
  * @param {expression} onTagAdded Expression to evaluate upon adding a new tag. The new tag is available as $tag.
  * @param {expression} onInvalidTag Expression to evaluate when a tag is invalid. The invalid tag is available as $tag.
+ * @param {expression} onTagRemoving A reference to a method in scope that will be invoked before removing a tag. This can be used to prevent a tag from being added. The tag is passed to the function for evaluation. This method must return either true or false. If false, the tag will not be removed.
  * @param {expression} onTagRemoved Expression to evaluate upon removing an existing tag. The removed tag is available as $tag.
  */
 tagsInput.directive('tagsInput', function($timeout, $document, tagsInputConfig, tiUtil) {
-    function TagList(options, events) {
+
+    function TagList(scope) {
```

Comment:
The TagList "class" should receive only what it needs to operate. It shouldn't receive the whole scope.